zkevm-circuits/src/evm_circuit/execution/extcodesize.rs

```diff
@@ -0,0 +1,326 @@
+            address,
+            tx_id,
+            reversion_info,
+            is_warm,
+            exists,
+            code_hash,
+            code_size,
+        }
+    }
+
+    fn assign_exec_step(
+        &self,
+        region: &mut CachedRegion<'_, '_, F>,
+        offset: usize,
+        block: &Block<F>,
+        tx: &Transaction,
+        call: &Call,
+        step: &ExecStep,
+    ) -> Result<(), Error> {
+        self.same_context.assign_exec_step(region, offset, step)?;
+
+        let address = block.rws[step.rw_indices[0]].stack_value().to_address();
+        let mut address_bytes = address.0;
+        address_bytes.reverse();
+        self.address.assign(region, offset, Some(address_bytes))?;
+
+        self.tx_id
+            .assign(region, offset, Value::known(F::from(tx.id as u64)))?;
+
+        self.reversion_info.assign(
+            region,
+            offset,
+            call.rw_counter_end_of_reversion,
+            call.is_persistent,
+        )?;
+
+        let (_, is_warm) = block.rws[step.rw_indices[4]].tx_access_list_value_pair();
+        self.is_warm
+            .assign(region, offset, Value::known(F::from(is_warm)))?;
+
+        let (exists, code_hash) = match block.rws[step.rw_indices[5]] {
+            Rw::Account {
+                field_tag: AccountFieldTag::CodeHash,
+                value,
+                ..
+            } => (true, value),
+            Rw::Account {
+                field_tag: AccountFieldTag::NonExisting,
+                ..
+            } => (false, 0.into()),
+            _ => unreachable!(),
+        };
+
+        let code_size = block.rws[step.rw_indices[6]].stack_value().as_u64();
+
+        self.exists
+            .assign(region, offset, Value::known(F::from(exists)))?;
+        self.code_hash.assign(
+            region,
+            offset,
+            Value::known(RandomLinearCombination::random_linear_combine(
+                code_hash.to_le_bytes(),
+                block.randomness,
+            )),
+        )?;
+        self.code_size
+            .assign(region, offset, Value::known(F::from(code_size)))?;
+
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use crate::evm_circuit::test::rand_bytes;
+    use crate::test_util::run_test_circuits;
+    use eth_types::geth_types::Account;
+    use eth_types::{address, bytecode, Address, Bytecode, Bytes, ToWord, Word};
+    use lazy_static::lazy_static;
+    use mock::TestContext;
+
+    lazy_static! {
+        static ref TEST_ADDRESS: Address = address!("0xaabbccddee000000000000000000000000000000");
+        static ref TEST_CODE: Bytes = Bytes::from([34, 54, 56]);
+        static ref TEST_ACCOUNT: Option<Account> = Some(Account {
+            address: *TEST_ADDRESS,
+            code: TEST_CODE.clone(),
+            ..Default::default()
+        });
+    }
+
+    #[test]
+    fn extcodesize_gadget_non_existing_account() {
+        test_root_ok(&None, false);
+        test_internal_ok(0x20, 0x00, &None, false);
```

Comment:
for root & internal call the extcodeside follows same handling , we can only keep one , only internal call test?

Reply:
Deleted `test_root_ok` and rename `test_internal_ok` to `test_ok`. And merged these test cases in zkevm-circuit as:
```
    #[test]
    fn test_extcodesize_gadget() {
        // Test for non existing account.
        test_ok(None, false);
        // Test for empty account.
        test_ok(Some(Account::default()), false);
        // Test for cold account.
        test_ok(TEST_ACCOUNT.clone(), false);
        // Test for warm account.
        test_ok(TEST_ACCOUNT.clone(), true);
    }
```